### PR TITLE
WIP: Low rank updates to the mass matrix

### PR DESCRIPTION
**Pull request by aseyboldt**
This is just a little weekend experiment for using a better mass matrix in nuts, and not really ready to merge.
But since I got a 2x speedup in the first model I tried, I thought maybe someone else might be interested.
The included notebook should explain at least very briefly.

### CODE REVIEW

File: pymc3/step_methods/hmc/quadpotential.py [hunk 1]
```diff
@@ -135,6 +139,165 @@ def energy(self, x):
...
+        samples[:] -= samples.mean(axis=1)[:, None]
+        self.stds = samples.std(axis=1)
+        samples[:] /= self.stds[:, None]
+        self.eigvals, self.eigvecs = self._find_eigs(samples, n_eigs, lam)
+        self.diag = 1 - (self.eigvecs ** 2 * self.eigvals).sum(axis=1)
+        self.n = len(samples)
+        #unexplained = self.n - self.eigvecs.sum()
+        #self.diag = np.ones(self.n) / self.n * unexplained
+
+    def __find_eigs(self, samples, n_eigs, lam):
```
Comment (ferrine) [inline pymc3/step_methods/hmc/quadpotential.py:267]: Why is it protected? (Begins with `__`)

**Comment by junpenglao** [comment]
Nice application of Ambikasaran's arXiv! I think it looks quite promising.

**Comment by twiecki** [comment]
I'll say, the sampling speed-up is very impressive.

File: pymc3/step_methods/hmc/quadpotential.py [hunk 2]
```diff
@@ -135,6 +139,165 @@ def energy(self, x):
...
+        return floatX(0.5) * tt.dot(x.T, self.velocity(x))
+
+
+class QuadPotentialTrace(object):
+    def __init__(self, model, trace, n_eigs, lam):
+        samples = np.array([model.dict_to_array(s) for s in trace]).T
+        samples[:] -= samples.mean(axis=1)[:, None]
+        self.stds = samples.std(axis=1)
+        samples[:] /= self.stds[:, None]
+        self.eigvals, self.eigvecs = self._find_eigs(samples, n_eigs, lam)
```
Comment (ferrine) [inline pymc3/step_methods/hmc/quadpotential.py:261]: Method defined below seems to be named differently

File: pymc3/step_methods/hmc/quadpotential.py [hunk 3]
```diff
@@ -71,7 +71,8 @@ def __init__(self, msg, idx):
         self.msg = msg
 
     def __str__(self):
-        return "Scaling is not positive definite. " + self.msg + ". Check indexes " + str(self.idx)
+        return "Scaling is not positive definite. %s. Check indexes %s" % \
+               (self.msg, str(self.idx))
 
 
 def isquadpotential(o):
```
Comment (ferrine) [inline pymc3/step_methods/hmc/quadpotential.py:78]: Why not refactoring this module and adding `BaseQuadPotential`?

**Comment by ferrine** [comment]
PR is great, notebook is really impressive

**Comment by aseyboldt** [comment]
@ferrine The implementation is a bit crappy at the moment and no tests at all. :-(
I want to figure out the math first. I think it is a bit strange that I can't find a lot of literature about preconditioning for hmc, most of what I can find that seems related is about using curvature information for optimization.

**Comment by twiecki** [comment]
@bob-carpenter just curious if you guys experimented with pre-condition of the mass matrix. we're seeing some interesting results here, but it's not very consistent.

**Comment by aseyboldt** [comment]
Link to the notebook: https://github.com/aseyboldt/pymc3/blob/7e33b38526fc550dcb27d1d00a0554c871dbff43/docs/source/notebooks/low-rank-mass.ipynb

**Comment by bob-carpenter** [comment]
The problem is always how to do the pre-conditioning.  If the posterior curvature doesn't vary, then it's easy and obviously a good thing to do if a Hessian-based estimate is possible and numerically stable.

The danger is that if the posterior curvature does vary and you precondition locally, say based on an initial point, you can wind up having to take some super small step sizes to make progress.  This can make it hard to find the typical set.  Examples include any kind of banana-shaped posterior (multiplicative effects) where the directions change or any kind of funnel shapes (hierarhchical model) where the scales and second derivatives change, but not the axis alignment.  

Stan remains conservative and tries to get to the typical set, then move around the typical set to estimate a diagonal mass matrix by default.  If the curvature varies, we'll keep stepping down the step size until we can move around.

Although we support the kind of dense mass matrix estimation you might expect (just a regularized sample-based estimate based on a moving window), we've found it to be rather unstable compared to diagonal mass matrices.  NONMEM just released a system where they allow block-diagonal mass matrices, breaking them down by patient for their PK/PD models with patient-level ("random") effects.

**Comment by twiecki** [comment]
That makes a lot of sense, thanks @bob-carpenter!

**Comment by twiecki** [comment]
Closing as it's more experimental and probably not ready to merge anytime soon.

**Pull request closed without merging**